drivers/lora/sx1276.c

```diff
@@ -0,0 +1,717 @@
+#define SX1276_REG_DIO_MAPPING1			0x40
+#define SX1276_REG_DIO_MAPPING2			0x41
+#define SX1276_REG_VERSION			0x42
+#define SX1276_REG_PA_DAC			0x4d
+
+#define SX1276_OPMODE_LONG_RANGE_MODE		BIT(7)
+#define SX1276_OPMODE_LOW_FREQUENCY_MODE_ON	BIT(3)
+#define SX1276_OPMODE_MODE_MASK			GENMASK(2, 0)
+#define SX1276_OPMODE_MODE_SLEEP		(0x0 << 0)
+#define SX1276_OPMODE_MODE_STDBY		(0x1 << 0)
+#define SX1276_OPMODE_MODE_TX			(0x3 << 0)
+#define SX1276_OPMODE_MODE_RXCONTINUOUS		(0x5 << 0)
+#define SX1276_OPMODE_MODE_RXSINGLE		(0x6 << 0)
+
+#define SX1276_MODEM_CONFIG1_BW_MASK		GENMASK(7, 4)
+#define SX1276_MODEM_CONFIG1_CR_MASK		GENMASK(3, 1)
+#define SX1276_MODEM_CONFIG1_IMP_HDR		BIT(0)
+#define SX1276_MODEM_CONFIG2_SF_MASK		GENMASK(7, 4)
+#define SX1276_PA_CONFIG_PA_BOOST		BIT(7)
+
+#define SX1276_DET_OPT_MASK			GENMASK(2, 0)
+#define SX1276_DET_OPT_SF7_SF12			0x03
+#define SX1276_DET_OPT_SF6			0x05
+#define SX1276_DET_THRES			GENMASK(7, 0)
+#define SX1276_DET_THRES_SF7_SF12		0x0a
+#define SX1276_DET_THRES_SF6			0x0c
+
+#define LORA_REG_IRQ_FLAGS_TX_DONE		BIT(3)
+#define LORA_REG_IRQ_FLAGS_RX_DONE		BIT(6)
+
+#define REG_DIO_MAPPING1_DIO0_MASK	GENMASK(7, 6)
+
+#define GPIO_DIO0_PIN		DT_INST_0_SEMTECH_SX1276_DIO_GPIOS_PIN
+#define GPIO_DIO0_CONTROLLER	DT_INST_0_SEMTECH_SX1276_DIO_GPIOS_CONTROLLER
+#define GPIO_RESET_PIN		DT_INST_0_SEMTECH_SX1276_RESET_GPIOS_PIN
+#define GPIO_CS_PIN		DT_INST_0_SEMTECH_SX1276_CS_GPIOS_PIN
+#define CLOCK_FREQ		DT_INST_0_SEMTECH_SX1276_CLOCK_FREQUENCY
+
+#define SX1276_PA_RFO				0
+#define SX1276_PA_BOOST				1
+
+struct sx1276_data {
+	struct device *dev;
+	struct device *spi;
+	struct spi_config spi_cfg;
+	struct device *dio0;
+	struct gpio_callback irq_cb;
+	struct k_sem data_sem;
+};
+
+struct _sx1276_power {
+	u8_t min;
+	u8_t max;
+};
+
+static struct _sx1276_power sx1276_power[2] = {
+	[SX1276_PA_RFO] = {
+		.min = 0,
+		.max = 14,
+	},
+	[SX1276_PA_BOOST] = {
+		.min = 2,
+		.max = 20,
+	},
+};
+static int sx1276_transceive(struct device *dev, u8_t reg,
+			     bool write, void *data, size_t length)
+{
+	struct sx1276_data *dev_data = dev->driver_data;
+	const struct spi_buf buf[2] = {
+		{
+			.buf = &reg,
+			.len = 1
+		},
+		{
+			.buf = data,
+			.len = length
+		}
+	};
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = 2,
+	};
+
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = buf,
+			.count = 2
+		};
+
+		return spi_transceive(dev_data->spi, &dev_data->spi_cfg,
+				      &tx, &rx);
+	}
+
+	return spi_write(dev_data->spi, &dev_data->spi_cfg, &tx);
+}
+
+int sx1276_read(struct device *dev, u8_t reg_addr, u8_t *data, u8_t len)
+{
+	return sx1276_transceive(dev, reg_addr, false, data, len);
+}
+
+int sx1276_write(struct device *dev, u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(dev, reg_addr | BIT(7), true, &byte, 1);
+}
+
+int sx1276_fifo_write(struct device *dev, u8_t *data, u32_t data_len)
+{
+	int ret, i;
+
+	for (i = 0; i < data_len; i++) {
```

Comment:
I think you don't need this loop. In the datasheet, there is the following statement:

`FIFO access: if the address byte corresponds to the address of the FIFO, then succeeding data byte will address the FIFO. The address is not automatically incremented but is memorized and does not need to be sent between each data byte. The NSS pin goes low at the beginning of the frame and stay low between each byte. It goes high only after the last byte transfer.`

So a burst transfer would bee sufficient here right?